### ISSUE TITLE
copy does not preserve compressed layers

### ISSUE THREAD

**Issue by kurt-google** [containers/skopeo#1406]
From the man page

```
--dest-compress bool-value
Compress tarball image layers when saving to directory using the 'dir' transport. (default is same compression type as source).
```

On skopeo 1.2.2 at least trying to copy a container with zstd compressed layers (from docker:// transport) rewrites these layers as gzip'ed data and does not preserve layer hashes. This seems incorrect based on the man page description.

It would also be nice if other transports also didn't recompress layers needlessly (when compression-algorithm is not set, and the transport supports the source format) for similar reasons of preserving layer data across skopeo calls. But that is more of a feature request I suppose.

**Comment by mtrmac** [comment]
Thanks for your report.

> On skopeo 1.2.2 at least trying to copy a container with zstd compressed layers (from docker:// transport) rewrites these layers as gzip'ed data and does not preserve layer hashes. This seems incorrect based on the man page description.

That’s surprising. Please attach at least a `--debug` log, and if possible a reproducer. Does the recent v1.4.0 behave the same way?

---

> It would also be nice if other transports also didn't recompress layers needlessly (when compression-algorithm is not set, and the transport supports the source format) for similar reasons of preserving layer data across skopeo calls. But that is more of a feature request I suppose.

That should, in general, be the case already. What behavior do you see that is not like that?

(Also there’s  #1378 , but that’s not quite the same thing.)

**Comment by mtrmac** [comment]
> That’s surprising

Ah, I see it; `--dest-compress-format` defaults to `gzip`, so `--dest-compress=true` means “compress using Gzip if not compressed, and re-compress using Gzip if compressed otherwise”. We should probably just fix that in c/image, and re-compress only on an explicit request.

So an immediate workaround for that case is to add `--dest-compress-format=zstd`, but that assumes you already know the format of the source (and that all layers use the same one).

---

> It would also be nice if other transports also didn't recompress layers needlessly

So is this the same issue, where copies to `docker://` or `oci`* default to re-compression using Gzip?

---

What are you ultimately trying to do? Preserve the original image? Compress it? Some combination?

(BTW if you want to use the `dir:` transform and preserve all hashes, just use `dir` without `--dest-compress` and without `--dest-decompress`.)

**Comment by kurt-google** [comment]
> > It would also be nice if other transports also didn't recompress layers needlessly

> So is this the same issue, where copies to docker:// or oci* default to re-compression using Gzip?

Right, I see the same layer rewrites as gzip when saving in oci/oci-archive.

> What are you ultimately trying to do? Preserve the original image? Compress it? Some combination?

Ultimately I was trying to use skopeo to validate images on repositories were being stored and served correctly as zstd layers.

**Comment by kurt-google** [comment]
> (Also there’s #1378 , but that’s not quite the same thing.)

That looks interesting as well. After noticing this issue I was not very excited that podman refused to implement compression algorithm options citing skopeo as the tool to use (https://github.com/containers/buildah/issues/3321#issuecomment-886181621) when it could not preserve formats, this is especially painful for very large images. Ensuring skopeo doesnt add excessive time to build chains would make it easier to accept the limitation at podman's level.

**Comment by mtrmac** [comment]
https://github.com/containers/image/pull/1333 should remove the unwanted recompression; can you verify that it works as expected, please?

#1378 is not too much work but a new feature, so not something I’ll immediately work on. PRs would be welcome, though.

**Comment by kurt-google** [comment]
Applying containers/image#1333 to the vendored dir and building skopeo master oci-archive works as I would expect without any flags now (performance is nicely improved during pull as well).

Thanks for the quick response and fix.

**Comment by mtrmac** [comment]
For the record, this was fixed in c/image v5.16.0, and will be fixed in the next Skopeo version (after Skopeo 1.4.1).

**Comment by rhatdan** [comment]
Since v5.16.0 is merged into skopeo main, I am closing.
